### PR TITLE
Feature/redis

### PR DESCRIPTION
**Pull request by 115100**

### CODE REVIEW

File: manager_redis.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+// Create a new policy to RedisManager
+func (m *RedisManager) Create(policy Policy) error {
+	payload, err := json.Marshal(policy)
+	if err != nil {
+		return err
+	}
+
+	keySet, err := m.db.HSetNX(m.redisPoliciesKey(), policy.GetID(), string(payload)).Result()
```
Comment (dim) [inline manager_redis.go]: Minor: bool vars in Go should follow the `isSomething` or `wasSomething` convention

File: manager_redis.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+		return err
+	}
+
+	return nil
+}
+
+func (m *RedisManager) getPolicyFromRedis(id string) (Policy, error) {
+	resp, err := m.db.HGet(m.redisPoliciesKey(), id).Bytes()
+	if err == redis.Nil {
+		return nil, errors.New("Not found")
```
Comment (dim) [inline manager_redis.go]: I think we should have a global var for it `var redisNotFound = errors.New("Not found")`

File: manager_redis.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	if !keySet {
+		return errors.New("Policy exists")
+	} else if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *RedisManager) getPolicyFromRedis(id string) (Policy, error) {
```
Comment (dim) [inline manager_redis.go]: Any need for this extra method? Can be just inlined in `Get`

File: manager_redis.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+// Delete removes a policy.
+func (m *RedisManager) Delete(id string) error {
+	return m.db.HDel(m.redisPoliciesKey(), id).Err()
+}
+
+// FindPoliciesForSubject finds all policies associated with the subject.
+func (m *RedisManager) FindPoliciesForSubject(subject string) (Policies, error) {
+	ps := Policies{}
```
Comment (dim) [inline manager_redis.go]: Declare `var ps Policies`, which is "weightless" while `Policies{}` is already allocating memory

File: manager_redis.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+			return nil, err
+		}
+
+		if ok, err := Match(p, p.GetSubjects(), subject); err != nil {
+			return nil, err
+		} else if !ok {
+			continue
+		}
+
+		ps = append(ps, p)
```
Comment (dim) [inline manager_redis.go:88]: `append` works with `nil` slices, so you can keep this as it is.

File: manager_redis.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+	}
+	if err := iter.Err(); err != nil {
+		return nil, errors.Wrap(err, "")
+	}
+
+	return ps, nil
+}
+
+func unmarshalPolicy(policy []byte) (Policy, error) {
```
Comment (dim) [inline manager_redis.go]: This should be prefixed with redis, i.e. `redisUnmarshalPolicy`

File: manager_redis.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+	p := new(DefaultPolicy)
```
Comment (dim) [inline manager_redis.go]: Don't (pre-)allocate, just `var p *DefaultPolicy` will do

File: manager_redis.go (hunk 3 above)
Comment (115100) [reply manager_redis.go]: Oops. Forgot to remove it in a refactor.

File: README.md [hunk 8]
```diff
@@ -455,6 +455,31 @@ func main() {
...
+	"github.com/ory-am/ladon"
+	"gopkg.in/redis.v5"
+)
+
+func main () {
+	db = redis.NewClient(&redis.Options{
+		Addr:     "localhost:6379",
+	})
+
+	if _, err := db.Ping().Result(); err != nil {
```
Comment (dim) [inline README.md]: `.Err()` will do, you are not reading the result.

**Review by dim — APPROVED** [review]

**Comment by 115100** [comment]
Squashed and pushed to new branch. Opening PR against upstream.

**Pull request closed without merging**